### PR TITLE
Track information about named common table expressions in the query

### PR DESCRIPTION
**Pull request by mlyublena**
## Description
<!---Describe your changes in detail-->
Collect information about named common table expressions defined either using the WITH clause, or by referencing a view.
For example:

```
presto:tpch> explain with tbl as (select * from lineitem where orderkey > 5), tbl2 as (select * from orders) select count(*) from tbl union all select max(orderkey) from tbl union all select count(*) from tbl2;                                                                                                                                                                                                                                                                                                                                          >
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[_col0] => [count_237:bigint]
         _col0 := count_237 (1:104)
     - LocalExchange[ROUND_ROBIN] () => [count_237:bigint]
         - Aggregate(FINAL) => [count:bigint]
                 count := "presto.default.count"((count_240)) (1:104)
...
 CTEInfo: [tbl2: 1, tbl: 2]
```
## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->
This will allow us to track information about queries that reference CTEs in their body.

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* ...
* ...

Hive Changes
* ...
* ...
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by jaystarshot** [comment]
Nice! We could also canonicalize and log the hash of the CTE along with some runtime stats in future which could help us identify heavy repeating CTEs

File: presto-analyzer/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 1]
```diff
@@ -180,6 +181,8 @@ public class Analysis
     // for materialized view analysis state detection, state is used to identify if materialized view has been expanded or in-process.
     private final Map<Table, MaterializedViewAnalysisState> materializedViewAnalysisStateMap = new HashMap<>();
 
+    private final HashSet<QualifiedObjectName> views = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (mlyublena) [reply presto-analyzer/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java:184]: good catch, I missed that! will add it

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/CTEInformationCollector.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import com.google.common.collect.ImmutableList;
+
+import java.util.HashMap;
+import java.util.List;
+
+public class CTEInformationCollector
+{
+    private final HashMap<String, CTEInformation> cteInformationMap = new HashMap<>();
+
+    public void addCTEReference(QualifiedName cteName, boolean isView)
```
[the remark this answers is not in the document]
Comment (mlyublena) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/CTEInformationCollector.java:27]: currently this is populated in the analyzer/relationplanner. I believe we don't need concurrency at the moment

**Review by feilong-liu — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/CTEInformationCollector.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        if (!cteInformationMap.containsKey(cteName.toString())) {
+            cteInformationMap.put(cteName.toString(), new CTEInformation(cteName.toString(), 0, isView));
+        }
+
+        cteInformationMap.get(cteName.toString()).incrementReferences();
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/CTEInformationCollector.java:33]: `cteInformationMap.putIfAbsent(...).incrementReferences()`

**Review by pranjalssh — APPROVED** [review]

**Pull request merged**